File: cancel-abandoned-payment-intents/node/cancel-abandoned.js
/**
 * Cancel abandoned Stripe PaymentIntents and their pending WooCommerce orders.
 * Never touches a real payment. Run on a schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/woocommerce/cancel-abandoned-payment-intents/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_HOURS = Number(process.env.LOOKBACK_HOURS || 168);
const THRESHOLD_HOURS = Number(process.env.THRESHOLD_HOURS || 12);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const ABANDONED = new Set(["requires_payment_method", "requires_confirmation"]);

export function isAbandoned(intent, ageHours, thresholdHours) {
  if (!ABANDONED.has(intent.status)) return false;
  if (intent.last_payment_error) return false;
  return ageHours >= thresholdHours;
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function* intentsWithAge(lookbackHours) {
  const since = Math.floor(Date.now() / 1000) - lookbackHours * 3600;
  for await (const intent of stripe.paymentIntents.list({ limit: 100, created: { gte: since } })) {
    if (intent.metadata.order_id) {
      const ageHours = (Date.now() / 1000 - intent.created) / 3600;
      yield { intent, ageHours };
    }
  }
}

export async function run() {
  let cancelled = 0;
  for await (const { intent, ageHours } of intentsWithAge(LOOKBACK_HOURS)) {
    if (!isAbandoned(intent, ageHours, THRESHOLD_HOURS)) continue;
    const orderId = intent.metadata.order_id;
    console.log(`Intent ${intent.id} (order ${orderId}) abandoned. ${DRY_RUN ? "would cancel" : "cancelling"}`);
    if (!DRY_RUN) {
      await stripe.paymentIntents.cancel(intent.id, { cancellation_reason: "abandoned" });
      const order = await woo(`/orders/${orderId}`);
      if (order && order.status === "pending") {
        await woo(`/orders/${orderId}`, { method: "PUT", body: JSON.stringify({ status: "cancelled" }) });
        await woo(`/orders/${orderId}/notes`, { method: "POST", body: JSON.stringify({ note: `Checkout was abandoned. Cancelled the Stripe PaymentIntent ${intent.id} and the order to release stock.` }) });
      }
    }
    cancelled++;
  }
  console.log(`Done. ${cancelled} abandoned intent(s) ${DRY_RUN ? "to cancel" : "cancelled"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: cancel-abandoned-payment-intents/python/cancel_abandoned.py
"""Cancel abandoned Stripe PaymentIntents and their pending WooCommerce orders.
Never touches a real payment. Run on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/woocommerce/cancel-abandoned-payment-intents/
"""
import os
import time
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("cancel_abandoned")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_HOURS = int(os.environ.get("LOOKBACK_HOURS", "168"))
THRESHOLD_HOURS = int(os.environ.get("THRESHOLD_HOURS", "12"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

ABANDONED = {"requires_payment_method", "requires_confirmation"}


def is_abandoned(intent, age_hours, threshold_hours):
    if intent.get("status") not in ABANDONED:
        return False
    if intent.get("last_payment_error"):
        return False
    return age_hours >= threshold_hours


def intents_with_age(lookback_hours):
    since = int(time.time()) - lookback_hours * 3600
    for intent in stripe.PaymentIntent.list(limit=100, created={"gte": since}).auto_paging_iter():
        if intent.metadata.get("order_id"):
            age_hours = (time.time() - intent["created"]) / 3600
            yield intent, age_hours


def get_order(order_id):
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def put_order(order_id, body):
    requests.put(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", json=body, auth=AUTH, timeout=30).raise_for_status()


def add_note(order_id, note):
    requests.post(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/notes",
                  json={"note": note}, auth=AUTH, timeout=30).raise_for_status()


def run():
    cancelled = 0
    for intent, age_hours in intents_with_age(LOOKBACK_HOURS):
        if not is_abandoned(intent, age_hours, THRESHOLD_HOURS):
            continue
        order_id = intent.metadata["order_id"]
        log.info("Intent %s (order %s) abandoned. %s", intent.id, order_id, "would cancel" if DRY_RUN else "cancelling")
        if not DRY_RUN:
            stripe.PaymentIntent.cancel(intent["id"], cancellation_reason="abandoned")
            order = get_order(order_id)
            if order and order["status"] == "pending":
                put_order(order_id, {"status": "cancelled"})
                add_note(order_id, f"Checkout was abandoned. Cancelled the Stripe PaymentIntent "
                                   f"{intent['id']} and the order to release stock.")
        cancelled += 1
    log.info("Done. %d abandoned intent(s) %s.", cancelled, "to cancel" if DRY_RUN else "cancelled")


if __name__ == "__main__":
    run()
